algorithms/src/Kokkos_Algo.hpp

```diff
@@ -0,0 +1,111 @@
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_ALGO_HPP
+#define KOKKOS_ALGO_HPP
+
+#include <Kokkos_Core.hpp>
+
+/// \file Kokkos_Algo.hpp
+/// \brief Kokkos counterparts for Standard C++ Library algorithms
+
+namespace Kokkos {
+namespace Experimental {
+
+template <class DataType, class... Properties>
+auto begin(Kokkos::View<DataType, Properties...>& v) -> decltype(v.data()) {
+  KOKKOS_EXPECTS(v.span_is_contiguous());
+  return v.data();
+}
+
+template <class DataType, class... Properties>
+auto begin(const Kokkos::View<DataType, Properties...>& v)
+    -> decltype(v.data()) {
+  KOKKOS_EXPECTS(v.span_is_contiguous());
+  return v.data();
+}
+
+template <class DataType, class... Properties>
+auto end(Kokkos::View<DataType, Properties...>& v) -> decltype(v.data()) {
+  KOKKOS_EXPECTS(v.span_is_contiguous());
+  return v.data() + v.size();
+}
+
+template <class DataType, class... Properties>
+auto end(const Kokkos::View<DataType, Properties...>& v) -> decltype(v.data()) {
+  KOKKOS_EXPECTS(v.span_is_contiguous());
+  return v.data() + v.size();
+}
+
+template <class PointerType, class FunctorType>
+FunctorType for_each(PointerType data, PointerType end, FunctorType functor) {
+  const auto numOfElements = end - data;
+  Kokkos::parallel_for(
+      numOfElements, KOKKOS_LAMBDA(const int i) {
+        auto element = data + i;
+        functor(*element);
+      });
+  return functor;
+}
+
+template <class DataType, class... Properties, class FunctorType>
+FunctorType for_each(Kokkos::View<DataType, Properties...> v,
+                     FunctorType functor) {
+  KOKKOS_EXPECTS(v.span_is_contiguous());
+  return for_each(v.data(), v.data() + v.size(), std::move(functor));
+}
+
+template <class PointerType, class SizeType, class FunctorType>
+PointerType for_each_n(PointerType data, SizeType n, FunctorType functor) {
```

Comment:
added:
```cpp
template <class DataType, class... Properties, class SizeType,
          class FunctorType>
void for_each_n(Kokkos::View<DataType, Properties...> v, SizeType n,
                FunctorType functor) {
```

Note: return type is `void` here, but it could (should?) be changed to `PointerType` to make it consistent with the other `for_each_n`.

Reply:
changed to return `IteratorType`